components/camel-ssh/src/main/java/org/apache/camel/component/ssh/ResourceBasedSSHKeyVerifier.java

```diff
@@ -0,0 +1,211 @@
+                    knownHostsResource);
+            List<String> possibleTokens = getKnownHostsFileTokensForSocketAddress(remoteAddress);
+            log.debug("Trying to mach PublicKey against provided known_hosts file");
+            PublicKey matchingKey = findKeyForServerToken(knownHostsInputStream, possibleTokens);
+            if (matchingKey != null) {
+                log.debug("Found PublicKey match for server");
+                boolean match = Arrays.areEqual(matchingKey.getEncoded(), serverKey.getEncoded());
+                return match;
+            }
+        } catch (IOException ioException) {
+            log.debug(String.format("Could not find known_hosts file %s", knownHostsResource), ioException);
+        } finally {
+            if (knownHostsInputStream != null) {
+                try {
+                    knownHostsInputStream.close();
+                } catch (IOException e) {
+                    // Ignore
+                }
+            }
+        }
+        if (failOnUnknownHost) {
+            log.warn("Could not find matching key for client session, connection will fail due to configuration");
+            return false;
+        } else {
+            log.warn(
+                    "Could not find matching key for client session, connection will continue anyway due to configuration");
+            return true;
+        }
+    }
+
+    private PublicKey findKeyForServerToken(InputStream knownHostsInputStream, List<String> possibleTokens)
+            throws IOException {
+        List<String> knowHostsLines = readInputStreamToStringList(knownHostsInputStream);
+
+        for (String s : knowHostsLines) {
+            String[] parts = s.split(" ");
+            if (parts.length != 3) {
+                log.warn("Found malformed entry in known_hosts file");
+                continue;
+            }
+            String entry = parts[0];
+            String key = parts[2];
+            for (String serverToken : possibleTokens) {
+                if (entry.contains(serverToken)) {
+                    try {
+                        return loadKey(key);
+                    } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
+                        log.warn(String.format("Could not load key for server token %s", entry), e);
+                    }
+                }
+            }
+        }
+        return null;
+    }
+
+    private List<String> readInputStreamToStringList(InputStream knownHostsInputStream) throws IOException {
+        List<String> returnList = new LinkedList<>();
+        String line;
+        BufferedReader bufferedReader = new BufferedReader(
+                new InputStreamReader(knownHostsInputStream, Charset.forName("UTF-8")));
+        while ((line = bufferedReader.readLine()) != null) {
+            returnList.add(line);
+        }
+        return returnList;
+    }
+
+    private List<String> getKnownHostsFileTokensForSocketAddress(SocketAddress remoteAddress) {

```

Comment:
this method is heavily based on String parsing. I don't know very well about known_hosts file's structure. Is this Ok with any file based know_hosts resource in the classpath? This makes code not very readable IMHO.